apisix/plugins/body-transformer.lua

```diff
@@ -0,0 +1,198 @@
+        input_format = { type = "string", enum = {"xml", "json"} },
+        template = { type = "string" },
+    },
+    required = {"template"},
+}
+
+local schema = {
+    type = "object",
+    properties = {
+        request = transform_schema,
+        response = transform_schema,
+    },
+}
+
+
+local _M = {
+    version  = 0.1,
+    priority = 1080,
+    name     = "body-transformer",
+    schema   = schema,
+}
+
+
+-- luacheck: ignore
+function string.escape_xml(s)
+    return s:gsub("&", "&amp;")
+        :gsub("<", "&lt;")
+        :gsub(">", "&gt;")
+        :gsub("'", "&apos;")
+        :gsub('"', "&quot;")
+end
+
+
+-- luacheck: ignore
+function string.escape_json(s)
+    return core.json.encode(s)
+end
+
+
+local function remove_namespace(tbl)
+    for k, v in pairs(tbl) do
+        if type(k) == "string" then
+            k = k:match(".*:(.*)")
+            if k then
+                tbl[k] = v
+            end
+            if type(v) == "table" then
+                remove_namespace(v)
+            end
+        end
+    end
+    return tbl
+end
+
+
+local decoders = {
+    xml = function(data)
+        local handler = xmlhandler:new()
+        local parser = xml2lua.parser(handler)
+        local ok, err = pcall(parser.parse, parser, data)
+        if ok then
+            return remove_namespace(handler.root)
+        else
+            return nil, err
+        end
+    end,
+    json = function(data)
+        return core.json.decode(data)
+    end,
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+local function transform(conf, body, typ, ctx)
+    local out = {_body = body}
+    if body then
+        local err
+        local format = conf[typ].input_format
+        if format then
+            out, err = decoders[format](body)
+            if not out then
+                err = str_format("%s body decode: %s", typ, err)
+                core.log.error(err, ", body=", body)
+                return nil, 400, err
+            end
+        end
+    end
+
+    local text = conf[typ].template
+    text = decode_base64(text) or text
+    local ok, render = pcall(template.compile, text)
+    if not ok then
+        local err = render
+        err = str_format("%s template compile: %s", typ, err)
+        core.log.error(err)
+        return nil, 500, err
+    end
+
+    out._ctx = ctx
+    ok, out = pcall(render, out)
+    if not ok then
+        local err = str_format("%s template rendering: %s", typ, out)
+        core.log.error(err)
+        return nil, 500, err
+    end
+
+    core.log.info(typ, " body transform output=", out)
+    return out
+end
+
+
+local function set_input_format(conf, typ, ct)
+    if conf[typ].input_format == nil and ct then
+        if ct:find("text/xml") then
+            conf[typ].input_format = "xml"
+        elseif ct:find("application/json") then
+            conf[typ].input_format = "json"
+        end
+    end
+end
+
+
+function _M.rewrite(conf, ctx)
```

Comment:
This plugin needs to use original content-type header before proxy-rewrite plugin overrrdes it.

Reply:
And this plugin is just doing rewrite job, just like proxy-rewrite plugin.

Reply:
We used to have a discussion to unify the "rewrite/access" methods in 2020.
Since this is not the only plugin that has inconsistent "rewrite/access" methods, and you insist on it, I am OK to accept it.